src/db/file.rs

```diff
@@ -225,7 +224,52 @@ fn file_list_to_json(file_list: Vec<(String, PathBuf)>) -> Result<Json> {
     Ok(file_list_json.to_json())
 }
 
+pub fn move_to_s3(conn: &Connection, n: usize) -> Result<()> {
+    let trans = try!(conn.transaction());
+    let client = s3_client().expect("configured s3");
+
+    let rows = try!(trans.query(
+            &format!("SELECT path, mime, content FROM files WHERE content != E'in-s3' LIMIT {}", n),
+            &[]));
+
+    let mut rt = ::tokio::runtime::current_thread::Runtime::new().unwrap();
+    let mut futures = Vec::new();
+    for row in &rows {
+        let path: String = row.get(0);
+        let mime: String = row.get(1);
+        let content: Vec<u8> = row.get(2);
+        let path_1 = path.clone();
+        futures.push(client.put_object(PutObjectRequest {
+            bucket: "rust-docs-rs".into(),
+            key: path.clone(),
+            body: Some(content.into()),
+            content_type: Some(mime),
+            ..Default::default()
+        }).map(move |_| {
+            path_1
+        }).map_err(move |e| {
+            panic!("failed to upload to {}: {:?}", path, e)
+        }));
+    }
+
+    use ::futures::future::Future;
+    match rt.block_on(::futures::future::join_all(futures)) {
+        Ok(paths) => {
+            let statement = trans.prepare("UPDATE files SET content = E'in-s3' WHERE path = $1")
+                .unwrap();
+            for path in paths {
+                statement.execute(&[&path]).unwrap();
+            }
+        }
+        Err(e) => {
+            panic!("results err: {:?}", e);
+        }
+    }
+
+    try!(trans.commit());
```

Comment:
In local testing I've confirmed that if anything goes wrong with the migration we will indeed abort the transaction here and nothing will get written to the database, i.e., we don't accidentally overwrite some of the files with "in-s3" despite not uploading them as such. It also means either all or nothing, but we're not doing more than 5000 at once so that's also fine.